cpu/avr8_common/thread_arch.c

```diff
@@ -266,31 +266,40 @@ void NORETURN avr8_enter_thread_mode(void)
     UNREACHABLE();
 }
 
+__attribute__((always_inline)) static inline void avr8_context_switch(void)
+{
+    avr8_context_save();
+    sched_run();
+    avr8_context_restore();
+}
+
 void thread_yield_higher(void)
 {
     if (irq_is_in() == 0) {
-        avr8_context_save();
-        sched_run();
-        avr8_context_restore();
-        __asm__ volatile ("ret");
+        avr8_context_switch();
     }
     else {
         sched_context_switch_request = 1;
     }
 }
 
-void avr8_exit_isr(void)
+__attribute__((always_inline)) static inline int avr8_nested_isr_allows_ctx_switch(void)
 {
-    avr8_state &= ~AVR8_STATE_FLAG_ISR;
-
-    /* Force access to avr8_state to take place */
-    __asm__ volatile ("" : : : "memory");
+#if !defined(CPU_ATXMEGA)
+    return GPIOR1 == 0;
+#else
+    return PMIC.STATUS <= PMIC_LOLVLEX_bm;
```

Comment:
I think this can cause issues. E.g. let's assume were in thread mode and a high priority thread is currently blocked waiting for an IRQ (e.g. via `mutex_lock()` on a locked mutex, and the ISR will call `mutex_unlock()`). Now a high priority IRQ comes in, which unblocks the high priority thread. But the high priority thread will not be scheduled when the ISR exists - instead it will return to the low priority thread.

So what is needed to test for is whether this will return to thread context, or to an ISR of lower prio.

Reply:
Yes, I believe it can happen with ATxmega.
Let me think a little bit more.

Reply:
I can imagine two paths:

1) implement prolog/epilog for ATxmega ISR body and define ISR as naked. This allows us add a cli as first instruction in the ISR block. In this case, we can mimics same ATmega behaviour.
2) Fix avrlibc ISR(xxxx, BLOCK) behaviour. This should add a "cli" as first line in the ISR block, right? But it is not working.

[1] allow us to remove avr8_enter_isr/avr8_exit_isr too.

Reply:
> This should add a "cli" as first line in the ISR block, right?

For ATmega it doesn't, as for ATmega implicitly by hardware IRQs are disabled upon ISR entry.

I guess the easiest would be to just use `GPIOR0` for ATxmega as well. But I'm honestly not really an expert with AVRs.

Reply:
This is how an ISR looks like for ATxmega. Note that cbi is atomic and works in their context. However, for switch context, it is very different situation because IRQ are always enabled for ATxmega. This means that a high prio ISR can execute between **13fa**	and **13fc**. That means, stack can be corrupted if when high prio exist it switches the context. We need add at **13dc** a **cli** instruction to mimic Atmega behavior.

This my understanding, not sure if I'm missing something.

```
000013dc <__vector_60>:
ISR(UART_0_TXC_ISR, ISR_BLOCK)
{

    13dc:	1f 92       	push	r1                                        # cli : instruction here solves the problem
    13de:	0f 92       	push	r0
    13e0:	0f b6       	in	r0, 0x3f	; 63
    13e2:	0f 92       	push	r0
    13e4:	11 24       	eor	r1, r1
    13e6:	08 b6       	in	r0, 0x38	; 56
    13e8:	0f 92       	push	r0
    13ea:	18 be       	out	0x38, r1	; 56
    13ec:	09 b6       	in	r0, 0x39	; 57
    13ee:	0f 92       	push	r0
    13f0:	19 be       	out	0x39, r1	; 57
    13f2:	0b b6       	in	r0, 0x3b	; 59
    13f4:	0f 92       	push	r0
    13f6:	1b be       	out	0x3b, r1	; 59
    13f8:	2f 93       	push	r18
    13fa:	3f 93       	push	r19
// high prio can run here!
    13fc:	4f 93       	push	r20
    13fe:	5f 93       	push	r21
    1400:	6f 93       	push	r22
    1402:	7f 93       	push	r23
    1404:	8f 93       	push	r24
    1406:	9f 93       	push	r25
    1408:	af 93       	push	r26
    140a:	bf 93       	push	r27
    140c:	ef 93       	push	r30
    140e:	ff 93       	push	r31

//    GPIOR0 &= ~AVR8_STATE_FLAG_UART_TX(uart);
    1410:	00 98       	cbi	0x00, 0	; 0

//    avr8_exit_isr();
    1412:	0e 94 48 0b 	call	0x1690	; 0x1690 <avr8_exit_isr>

    1416:	ff 91       	pop	r31
    1418:	ef 91       	pop	r30
    141a:	bf 91       	pop	r27
    141c:	af 91       	pop	r26
    141e:	9f 91       	pop	r25
    1420:	8f 91       	pop	r24
    1422:	7f 91       	pop	r23
    1424:	6f 91       	pop	r22
    1426:	5f 91       	pop	r21
    1428:	4f 91       	pop	r20
    142a:	3f 91       	pop	r19
    142c:	2f 91       	pop	r18
    142e:	0f 90       	pop	r0
    1430:	0b be       	out	0x3b, r0	; 59
    1432:	0f 90       	pop	r0
    1434:	09 be       	out	0x39, r0	; 57
    1436:	0f 90       	pop	r0
    1438:	08 be       	out	0x38, r0	; 56
    143a:	0f 90       	pop	r0
    143c:	0f be       	out	0x3f, r0	; 63
    143e:	0f 90       	pop	r0
    1440:	1f 90       	pop	r1
    1442:	18 95       	reti
```

Reply:
I don't see an issue with an IRQ interrupting a low prio ISR while pushing the registers to the stack: The higher prio ISR will also push the contents of every register it might modify to the stack and pop their values again once it completes. Hence, the low prio ISR will continue to execute the very instruction it was interrupted before with the exact state of the registers and stack. Except for the lost time, the low prio ISR wouldn't even notice it was ever interrupted.

Reply:
I think I can fix the issue. Let's see if I can send something to in the next days.